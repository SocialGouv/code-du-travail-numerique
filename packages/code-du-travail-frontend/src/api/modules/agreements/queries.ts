import { SOURCES } from "@socialgouv/cdtn-sources";

export const getAllAgreementsWithContributions = () => {
  return {
    _source: ["title", "shortTitle", "description", "url", "slug", "source"],
    query: {
      bool: {
        filter: [
          { term: { source: SOURCES.CCN } },
          { term: { isPublished: true } },
          { term: { contributions: true } },
        ],
      },
    },
    size: 200,
  };
};

export function getAgreementsBySlugs(slugs: string[]) {
  return {
    _source: [
      "title",
      "shortTitle",
      "description",
      "url",
      "slug",
      "breadcrumbs",
      "source",
      "cdtnId",
    ],
    query: {
      bool: {
        filter: [
          { term: { source: SOURCES.CCN } },
          { term: { isPublished: true } },
          { terms: { slug: slugs } },
        ],
      },
    },
    size: 100,
  };
}

export function getAgreementsByIds(ids: string[]) {
  return {
    _source: [
      "title",
      "shortTitle",
      "description",
      "url",
      "slug",
      "breadcrumbs",
      "source",
      "cdtnId",
    ],
    query: {
      bool: {
        filter: [
          { term: { source: SOURCES.CCN } },
          { term: { isPublished: true } },
          { terms: { cdtnId: ids } },
        ],
      },
    },
    size: 100,
  };
}

export const getAgreementBySlugBody = (slug: string) => {
  return {
    query: {
      bool: {
        filter: [
          { term: { slug } },
          { term: { source: SOURCES.CCN } },
          { term: { isPublished: true } },
        ],
      },
    },
    size: 1,
  };
};
