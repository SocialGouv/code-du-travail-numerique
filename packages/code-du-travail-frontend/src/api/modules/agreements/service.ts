import { SearchResponse, ElasticSearchItem, Agreement } from "cdtn-types";
import { elasticsearchClient, elasticIndex, NotFoundError } from "../../utils";
import {
  getAllAgreementsWithContributions,
  getAgreementsBySlugs,
  getAgreementBySlugBody,
  getAgreementsByIds,
} from "./queries";

export const getAllAgreements = async (): Promise<Agreement[]> => {
  const body = getAllAgreementsWithContributions();

  const response = await elasticsearchClient.search<SearchResponse<Agreement>>({
    body,
    index: elasticIndex,
  });

  return response.body.hits.hits
    .map(({ _source }) => _source)
    .sort(orderByAlphaAndMetalurgieLast);
};

export const getBySlugsAgreements = async (
  slugs: string[]
): Promise<ElasticSearchItem[]> => {
  const body = getAgreementsBySlugs(slugs);
  const response = await elasticsearchClient.search({
    body,
    index: elasticIndex,
  });
  return response.body.hits.total.value > 0
    ? response.body.hits.hits.map(({ _source }) => _source)
    : [];
};

export const getByIdsAgreements = async (
  ids: string[]
): Promise<ElasticSearchItem[]> => {
  const body = getAgreementsByIds(ids);
  const response = await elasticsearchClient.search({
    body,
    index: elasticIndex,
  });
  return response.body.hits.total.value > 0
    ? response.body.hits.hits.map(({ _source }) => _source)
    : [];
};

export const getBySlugAgreements = async (slug: string) => {
  const body = await getAgreementBySlugBody(slug);

  const response = await elasticsearchClient.search({
    body,
    index: elasticIndex,
  });
  if (response.body.hits.total.value === 0) {
    throw new NotFoundError({
      message: `Agreement not found, no agreement match ${slug}y`,
      name: "AGREEMENT_NOT_FOUND",
      cause: null,
    });
  }

  return { ...response.body.hits.hits[0]._source };
};

const orderByAlphaAndMetalurgieLast = (a, b) => {
  if (a.url && !b.url) {
    return -1;
  }
  if (!a.url && b.url) {
    return 1;
  }
  if (a.title < b.title) {
    return -1;
  }
  if (a.title > b.title) {
    return 1;
  }

  return 0;
};
