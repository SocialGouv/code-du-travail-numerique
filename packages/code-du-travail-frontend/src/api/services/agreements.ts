import {
  Agreement,
  elasticIndex,
  elasticsearchClient,
  NotFoundError,
  SearchResponse,
} from "../utils";
import {
  getAgreementBySlugBody,
  getAllAgreementsWithContributions,
} from "./queries";

export class AgreementsService {
  public async getAll(): Promise<Agreement[]> {
    const body = getAllAgreementsWithContributions();

    const response = await elasticsearchClient.search<
      SearchResponse<Agreement>
    >({
      body,
      index: elasticIndex,
    });

    return response.body.hits.hits
      .map(({ _source }) => _source)
      .sort(orderByAlphaAndMetalurgieLast);
  }

  public async getBySlug(slug: string) {
    const body = await getAgreementBySlugBody(slug);

    const response = await elasticsearchClient.search({
      body,
      index: elasticIndex,
    });
    if (response.body.hits.total.value === 0) {
      throw new NotFoundError({
        message: `Agreement not found, no agreement match ${slug}y`,
        name: "AGREEMENT_NOT_FOUND",
        cause: null,
      });
    }

    return { ...response.body.hits.hits[0]._source };
  }
}

const orderByAlphaAndMetalurgieLast = (a, b) => {
  if (a.url && !b.url) {
    return -1;
  }
  if (!a.url && b.url) {
    return 1;
  }
  if (a.title < b.title) {
    return -1;
  }
  if (a.title > b.title) {
    return 1;
  }

  return 0;
};
