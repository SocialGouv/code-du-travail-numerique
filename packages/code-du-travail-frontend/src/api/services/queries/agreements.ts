import { SOURCES } from "@socialgouv/cdtn-sources";

export const getAllAgreementsWithContributions = () => {
  return {
    _source: ["title", "shortTitle", "description", "url", "slug", "source"],
    query: {
      bool: {
        filter: [
          { term: { source: SOURCES.CCN } },
          { term: { isPublished: true } },
          { term: { contributions: true } },
        ],
      },
    },
    size: 200,
  };
};
export const getAgreementBySlugBody = (slug: string) => {
  return {
    query: {
      bool: {
        filter: [
          { term: { slug } },
          { term: { source: SOURCES.CCN } },
          { term: { isPublished: true } },
        ],
      },
    },
    size: 1,
  };
};
