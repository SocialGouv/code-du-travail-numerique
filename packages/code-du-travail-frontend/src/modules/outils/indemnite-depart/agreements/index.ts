export * from "./types";
export { default as AgreementsInjector } from "./AgreementsInjector";
export { default as validatorAgreement } from "./validator";
export * from "./messages";
