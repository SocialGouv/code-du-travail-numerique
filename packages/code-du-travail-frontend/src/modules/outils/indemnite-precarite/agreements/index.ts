// Questions humanisées pour les critères
export const questions = {
  cddType: "Quel est le type de CDD ?",
  hasCdiProposal:
    "À la fin du CDD, le salarié a-t-il reçu une proposition de CDI ?",
  hasCdiRenewal:
    "À la fin du CDD, le salarié a-t-il été immédiatement embauché en CDI ?",
  hasEquivalentCdiRenewal:
    "À la fin du CDD, le salarié a-t-il été immédiatement embauché en CDI, sans interruption, sur un même poste ou sur un poste différent ?",
};

// Labels pour les critères
export const labels = {
  ccn: "convention collective",
  cddType: "type de cdd",
  hasCdiProposal: "proposition d'un cdi",
  hasCdiRenewal: "renouvellement en cdi",
  hasEquivalentCdiRenewal: "renouvellement en cdi équivalent",
};

// Ordre des critères pour les questions dynamiques
export const criteriaOrder = [
  "cddType",
  "hasCdiProposal",
  "hasCdiRenewal",
  "hasEquivalentCdiRenewal",
];

// Validation spécifique pour certaines conventions collectives
export function validateSpecificConventions(
  values: Record<string, any>
): Record<string, any> {
  const errors: Record<string, any> = {};

  // Convention collective 3127
  if (
    values?.ccn?.selected?.num === 3127 &&
    values?.criteria?.cddType ===
      "CDD dit de « mission ponctuelle ou occasionnelle »" &&
    values?.criteria?.hasEquivalentCdiRenewal === "oui"
  ) {
    errors.criteria = {
      hasEquivalentCdiRenewal:
        "Selon votre convention collective, lorsque le contrat de mission ponctuelle est transformé en CDI pour un poste et une durée équivalents, le salarié n'a pas le droit à une prime d'intervention.",
    };
  }

  // Convention collective 1486
  if (
    values?.ccn?.selected?.num === 1486 &&
    values?.criteria?.cddType ===
      "Contrat d'intervention dans le secteur d'activité d'organisation des foires, salons et congrès" &&
    values?.criteria?.hasCdiProposal === "oui"
  ) {
    errors.criteria = {
      hasCdiProposal:
        "Selon votre convention collective, le salarié en contrat d'intervention qui, à l'issue de son contrat, a reçu une proposition d'un CDI, n'a pas le droit à une prime d'intervention.",
    };
  }

  // Convention collective 2511
  if (
    values?.ccn?.selected?.num === 2511 &&
    values?.criteria?.cddType ===
      "CDD d'usage appelé contrat «d'intervention»" &&
    values?.criteria?.hasCdiRenewal === "oui"
  ) {
    errors.criteria = {
      hasCdiRenewal:
        "Selon votre convention collective, lorsque le contrat d'intervention est transformé en CDI, le salarié n'a pas le droit à une prime d'intervention.",
    };
  }

  // Convention collective 1516
  if (
    values?.ccn?.selected?.num === 1516 &&
    values?.criteria?.cddType === "CDD d'usage" &&
    values?.criteria?.hasCdiRenewal === "oui"
  ) {
    errors.criteria = {
      hasCdiRenewal: `Selon votre convention collective, le salarié en contrat d'usage qui, à l'issu de son contrat, poursuit par un CDI, n'a pas le droit à une indemnité dite "d'usage".`,
    };
  }

  return errors;
}
