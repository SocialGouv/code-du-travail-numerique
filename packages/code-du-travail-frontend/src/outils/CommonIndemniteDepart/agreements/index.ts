export * from "./types";
export * from "./index";
export { default as AgreementsInjector } from "./AgreementsInjector";
export { default as validatorAgreement } from "./validator";
export * from "./ui-customizations";
