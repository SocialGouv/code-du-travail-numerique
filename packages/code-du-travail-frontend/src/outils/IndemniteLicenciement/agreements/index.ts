export * from "./types";
export * from "./index";
export { default as AgreementsInjector } from "./AgreementsInjector";
export * from "./validator";
