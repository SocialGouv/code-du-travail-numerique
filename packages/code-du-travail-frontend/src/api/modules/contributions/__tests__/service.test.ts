import {
  getAllContributionsGroupByQuestion,
  getByIdsContributions,
  getGenericContributionsGroupByThemes,
} from "../service";
import { getAllAgreements } from "../../agreements";

describe("Contributions", () => {
  it("getGenericContributions", async () => {
    const result = await getGenericContributionsGroupByThemes();
    expect(result).toMatchSnapshot();
  });

  it("getByIdsContributions", async () => {
    const result = await getByIdsContributions(["eba7a4592f"]);
    expect(result).toMatchSnapshot();
  });

  it("getAllContributions", async () => {
    const agreements = await getAllAgreements();
    const result = await getAllContributionsGroupByQuestion(agreements);
    expect(result).toMatchSnapshot();
  });
});
