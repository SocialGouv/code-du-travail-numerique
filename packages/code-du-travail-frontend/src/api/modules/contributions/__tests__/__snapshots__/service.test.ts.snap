// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contributions getAllContributions 1`] = `
[
  {
    "breadcrumbs": [
      {
        "label": "Congés et repos",
        "position": 3,
        "slug": "/themes/conges-et-repos",
      },
      {
        "label": "Congés",
        "position": 1,
        "slug": "/themes/conges",
      },
      {
        "label": "Congés pour événement familial",
        "position": 3,
        "slug": "/themes/conges-pour-evenement-familial",
      },
    ],
    "description": "Le code du travail prévoit que le salarié bénéficie de congés à l’occasion de certains événements familiaux. Ils n’entrainent aucune diminution de la rémunération…",
    "slug": "les-conges-pour-evenements-familiaux",
    "source": "contributions",
    "title": "Les congés pour événements familiaux",
  },
  {
    "breadcrumbs": [
      {
        "label": "Départ de l’entreprise",
        "position": 7,
        "slug": "/themes/depart-de-lentreprise",
      },
      {
        "label": "Autres départs",
        "position": 6,
        "slug": "/themes/autres-departs",
      },
    ],
    "description": "Le décès de l’employeur a des conséquences différentes sur le contrat de travail, selon que l’employeur est une entreprise individuelle ou un employeur…",
    "slug": "quelles-sont-les-consequences-du-deces-de-lemployeur-sur-le-contrat-de-travail",
    "source": "contributions",
    "title": "Quelles sont les conséquences du décès de l’employeur sur le contrat de travail ?",
  },
  {
    "breadcrumbs": [
      {
        "label": "Départ de l’entreprise",
        "position": 7,
        "slug": "/themes/depart-de-lentreprise",
      },
      {
        "label": "Autres départs",
        "position": 6,
        "slug": "/themes/autres-departs",
      },
    ],
    "description": "Le décès de l’employeur a des conséquences différentes sur le contrat de travail, selon que l’employeur est une entreprise individuelle ou un employeur…",
    "slug": "quand-le-salarie-a-t-il-droit-a-une-prime-danciennete-quel-est-son-montant",
    "source": "contributions",
    "title": "Quand le salarié a-t-il droit à une prime d’ancienneté ? Quel est son montant ?",
  },
]
`;

exports[`Contributions getByIdsContributions 1`] = `
[
  {
    "breadcrumbs": [
      {
        "label": "Congés et repos",
        "position": 3,
        "slug": "/themes/conges-et-repos",
      },
      {
        "label": "Congés",
        "position": 1,
        "slug": "/themes/conges",
      },
      {
        "label": "Congés pour événement familial",
        "position": 3,
        "slug": "/themes/conges-pour-evenement-familial",
      },
    ],
    "cdtnId": "eba7a4592f",
    "description": "Le code du travail prévoit que le salarié bénéficie de congés à l’occasion de certains événements familiaux. Ils n’entrainent aucune diminution de la rémunération…",
    "slug": "les-conges-pour-evenements-familiaux",
    "source": "contributions",
    "title": "Les congés pour événements familiaux",
  },
]
`;

exports[`Contributions getBySlugsContributions 1`] = `
[
  {
    "breadcrumbs": [
      {
        "label": "Congés et repos",
        "position": 3,
        "slug": "/themes/conges-et-repos",
      },
      {
        "label": "Congés",
        "position": 1,
        "slug": "/themes/conges",
      },
      {
        "label": "Congés pour événement familial",
        "position": 3,
        "slug": "/themes/conges-pour-evenement-familial",
      },
    ],
    "cdtnId": "eba7a4592f",
    "description": "Le code du travail prévoit que le salarié bénéficie de congés à l’occasion de certains événements familiaux. Ils n’entrainent aucune diminution de la rémunération…",
    "slug": "les-conges-pour-evenements-familiaux",
    "source": "contributions",
    "title": "Les congés pour événements familiaux",
  },
]
`;

exports[`Contributions getGenericContributions 1`] = `
{
  "Congés et repos": [
    {
      "breadcrumbs": [
        {
          "label": "Congés et repos",
          "position": 3,
          "slug": "/themes/conges-et-repos",
        },
        {
          "label": "Congés",
          "position": 1,
          "slug": "/themes/conges",
        },
        {
          "label": "Congés pour événement familial",
          "position": 3,
          "slug": "/themes/conges-pour-evenement-familial",
        },
      ],
      "description": "Le code du travail prévoit que le salarié bénéficie de congés à l’occasion de certains événements familiaux. Ils n’entrainent aucune diminution de la rémunération…",
      "slug": "les-conges-pour-evenements-familiaux",
      "source": "contributions",
      "theme": "Congés et repos",
      "title": "Les congés pour événements familiaux",
    },
  ],
  "Départ de l’entreprise": [
    {
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 7,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Autres départs",
          "position": 6,
          "slug": "/themes/autres-departs",
        },
      ],
      "description": "Le décès de l’employeur a des conséquences différentes sur le contrat de travail, selon que l’employeur est une entreprise individuelle ou un employeur…",
      "slug": "quelles-sont-les-consequences-du-deces-de-lemployeur-sur-le-contrat-de-travail",
      "source": "contributions",
      "theme": "Départ de l’entreprise",
      "title": "Quelles sont les conséquences du décès de l’employeur sur le contrat de travail ?",
    },
    {
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 7,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Autres départs",
          "position": 6,
          "slug": "/themes/autres-departs",
        },
      ],
      "description": "Le décès de l’employeur a des conséquences différentes sur le contrat de travail, selon que l’employeur est une entreprise individuelle ou un employeur…",
      "slug": "quand-le-salarie-a-t-il-droit-a-une-prime-danciennete-quel-est-son-montant",
      "source": "contributions",
      "theme": "Départ de l’entreprise",
      "title": "Quand le salarié a-t-il droit à une prime d’ancienneté ? Quel est son montant ?",
    },
  ],
}
`;
